include/openssl/core_names.h

```diff
@@ -185,6 +188,13 @@ extern "C" {
 #define OSSL_PKEY_PARAM_DSA_PUB_KEY  "pub"
 #define OSSL_PKEY_PARAM_DSA_PRIV_KEY "priv"
 
+/* Elliptic Curve Keys */
+#define OSSL_PKEY_PARAM_EC_PUB_KEY   "pub"
```

Comment:
When you rebase these should disappear (Just use OSSL_PKEY_PARAM_PUB_KEY instead)

Reply:
This was discussed previously in this PR, I would prefer not to use bignum notation for separate x and y coordinates.

At least w.r.t. point encoding we have a very clear and straightforward way of representing a point as an octet string, I don't see a need to have our own alternative way of representing a point if there is no real reason for it: this has already opened us to potential misuse of the API and the risk of increased attack surface or interoperability issues (see #10366)

Reply:
I see that #10964 removed the separate priv/pub names for the different keys in favor of `OSSL_PKEY_PARAM_{PRIV,PUB}_KEY`

https://github.com/openssl/openssl/pull/10964/files#diff-f6a04511bcc584ff44f0c7266f2a7a5f

I did not have time to review that PR, but I am not sure it was a good idea.

It's true they all mapped to the same constant string as a name, but it was useful to have separate semantic types, because for different keys that OSSL_PARAM has different types: sometimes it is a `BIGNUM`, others an octet string!

@mattcaswell @levitte @slontis what are your thoughts on this?

Reply:
Know your provider.

The design is that a name (string) is just a name, it doesn't carry any type info. The associated type info should be documented on a per provider basis (and we're really sloppy with that... EVP_MAC and EVP_KDF paved the way, but...), and we have recently started to demand that there be settable and gettables that specify the associated types as well, on a per provider basis.

So, know your provider

Reply:
Well, from the perspective of an external provider developer, at the moment this is not really "know your provider" at the moment, but more "know what the mainlined providers/libcrypto expect".

Bu anyway, I will follow this advice, could you also suggest where and how I should document the semantics for the EC specific `OSSL_PARAM`s?

Reply:
I'm out and about, so not in a position to give you enormous detail. The pattern we followed for EVP_MAC and EVP_KDF is "doc/man7/{type}-{name}.pod", where {type} is the method type name, and {name} is the algorithm name (or for EVP_KEYMGMT, the key type name). So for example, I would expect to see a EVP_KEYMGMT-EC.pod.

(since we have many operations using the keytype name as algorithm name, we will have quite a few documents for EC... not sure how optimal that is)

Another option is to have one pod file in doc/man7 named like the keytype/operation algorithm, and describe all its aspects there.

How documentation should be written hasn't been settled, as you can see. Your input is welcome! However, I think that doc/man7 is the location to place them in.

Reply:
Thanks @levitte 

I would go with @slontis previous suggestion of separating that part of the documentation from this PR (and it seems sensible so that we can give the same treatment to DH, DSA, RSA, EC, ECX).

Would that be acceptable to you as a reviewer?

Reply:
Sure

Reply:
> This was discussed previously in this PR, I would prefer not to use bignum notation for separate x and y coordinates.

We  should support both formats... Here is some test data for KAS CAVS

The x,y,d files all have the same form (BIGNUMS is the way it normally passes these values around as BIGNUMS).
dsCAVS = 33d04383a9895603d2d7fe6b016fe459cc0d9a246c861b2edc4b4d3543e11bad
QsCAVSx = 1b9367551c559f63d122a4d8d10a606d02a57757c8a347733a6a082839bdc9d2
QsCAVSy = 80ece9a70829712fc95682ee9a850f6d7f595f8cd1960bdf293e4907883f9a29
dsIUT = 03530e2854a35c57211d09183a3679f62c6f63c6f517798d151fa90f83f98e18
QsIUTx = 1f72bd2a3eeb6c76e55d697524bf2f5b96b291620635ccb24b311b0c6f069f86
QsIUTy = cfc8acd54f4d77f3704a8f049ab103c7ebd5947861ab780c4a2d6bf32f2e4abc

I don't want to externally call a EC_  API (that should probably be deprecated) to convert this into some other format.. (This is not part of PKEY).
I just want to set the raw values into the key..
Whether I do this externally to convert it to octets before passing in, doesnt seem to vary the attack surface.

the set would look like
if locate(PARAM_PUBLIC, OCTET STRING) {
 .. set the point
} else if locate (PARAM_PUBLIC_X, UNSIGNED_INT) && locate (PARAM_PUBLIC_Y, UNSIGNED_INT)  {
  call internal EC method to convert x,y to point.
}

Reply:
Well.. to me it seems the NIST data is not conforming to anything and not even defining rigorously how the affine coordinates are being represented...

The reference standard on how to represent a field element as an octet string is SECG SEC 1 v2.0 Paragrpah 2.3.5 and is well defined for coordinates on a prime field or a binary field.

For the NIST data it's not really defined rigorously how to interpret those hex strings, and in the CAVP FAQ about ECDSA (coudln't find a relevant section for ECC CDH) says:

> ## 12 ECDSA FAQ
>
> ### ECDSA.1 For ECDSA PKV validation testing, how are the `Qx` and `Qy` values represented? What is the significance of their representation?
>
> All values should be thought of as hexadecimal numbers. To determine whether a value
or a point is valid, convert it to a number; do not rely on the format (e.g., leading zeros,
number of hexadecimal symbols, etc.)

This doesn't really say how to convert the hexadecimal strings to "numbers", endianess, nor how it would map to a binary field coordinate.

Moreover sending separate `x` and `y` pairs creates a problem for representing the point at infinity.

My opinion is that any code that makes assumptions on how to read the NIST data and convert it to valid point coordinates should stay confined to the portion of code where those tests are being parsed/stored.

Tainting with ambiguity the interface between `libcrypto <-> providers` and `providers <-> providers` because of the ill defined format used by these tests does not seem the right choice.

Reply:
I would prefer if your proposal was discussed in a separate PR where we could evaluate the benefits of having 2 ways of defining an EC public key with the costs of the ambiguity that the NIST format brings.

Reply:
You don't necessarily need to call an EC function to generate octet strings out of the NIST format, just concatenate in a buffer (hex notation):

`04` | `<padded_x>` | `<padded_y>`

where `<padded_{x,y}>` are zero padded octet strings for each of the coordinates, whose length in bytes is set to the number of bytes required to write the field order

Reply:
Ok this doesnt need to be solved for this PR..

Reply:
I understand the format : I am just not sure it should be pushed back onto the user.

Reply:
Not sure what you are trying to say for the infinite point case - what would be the octet string for that?

Reply:
Point at infinity should be encoded as the octet string `00` according to SECG.

This is because it is a special point so its "affine coordinates" are not defined, see for example https://crypto.stackexchange.com/questions/6464/how-to-represent-point-at-infinity-in-affine-coordinate

Now, it's a given that point-at-infinity will never be a valid EC public key, but it is an example of the subtle things that could happen by making our own homemade representation of curve points.

> I understand the format : I am just not sure it should be pushed back onto the user.

To me it seems that your proposal would be pushing the ill-defined NIST format onto the user (each provider developer).

Also having to support two formats means that to be interoperable every implementation must  support both representations.

If you are talking about the user of `libcrypto` as in the developer of third party applications or the user of one of the openssl apps, I don't know how things are going to change there.

Are we deprecating any `EC_POINT_*`, `EC_GROUP_*` function already? I am not aware of  a replacement in EVP for the functions to manipulate the EC objects (or to pick apart a DSA/RSA/DH key or its domain parameters) so I don't think those can be put on the path for deprecation!
Unless we are deprecating those, the user can use `EC_POINT_set_affine_coordinates()`.

Reply:
There is no plan to deprecate the EC_* functions (just the ECDSA_* and ECDH_* functions)